crates/uplc/src/machine/runtime.rs

```diff
@@ -1454,6 +1497,268 @@ impl DefaultFunction {
+                        .map(|b| match b {
+                            itertools::EitherOrBoth::Both(left_byte, right_byte) => {
+                                left_byte | right_byte
+                            }
+                            // Shorter is appended with 00 bytes that when or-ed produce the other bytestring
+                            itertools::EitherOrBoth::Left(byte)
+                            | itertools::EitherOrBoth::Right(byte) => *byte,
+                        })
+                        .collect_vec()
+                } else {
+                    bytes1
+                        .iter()
+                        .zip(bytes2)
+                        .map(|(b1, b2)| b1 | b2)
+                        .collect_vec()
+                };
+
+                Ok(Value::byte_string(bytes_result))
+            }
+            DefaultFunction::XorByteString => {
+                let should_pad = args[0].unwrap_bool()?;
+                let bytes1 = args[1].unwrap_byte_string()?;
+                let bytes2 = args[2].unwrap_byte_string()?;
+
+                let bytes_result = if *should_pad {
+                    bytes1
+                        .iter()
+                        .zip_longest(bytes2)
+                        .map(|b| match b {
+                            itertools::EitherOrBoth::Both(left_byte, right_byte) => {
+                                left_byte ^ right_byte
+                            }
+                            // Shorter is appended with 00 bytes that when xor-ed produce the other bytestring
+                            itertools::EitherOrBoth::Left(byte)
+                            | itertools::EitherOrBoth::Right(byte) => *byte,
+                        })
+                        .collect_vec()
+                } else {
+                    bytes1
+                        .iter()
+                        .zip(bytes2)
+                        .map(|(b1, b2)| b1 ^ b2)
+                        .collect_vec()
+                };
+
+                Ok(Value::byte_string(bytes_result))
+            }
+            DefaultFunction::ComplementByteString => {
+                let bytes = args[0].unwrap_byte_string()?;
+
+                let result = bytes.iter().map(|b| b ^ 255).collect_vec();
+
+                Ok(Value::byte_string(result))
+            }
+            DefaultFunction::ReadBit => {
+                let bytes = args[0].unwrap_byte_string()?;
+                let bit_index = args[1].unwrap_integer()?;
+
+                // This ensures there is at least one byte in bytes
+                if *bit_index < 0.into() || *bit_index >= (bytes.len() * 8).into() {
+                    return Err(Error::ReadBitOutOfBounds);
+                }
+
+                let (byte_index, bit_offset) = bit_index.div_rem(&8.into());
+
+                let bit_offset = usize::try_from(bit_offset).unwrap();
+
+                let flipped_index = bytes.len() - 1 - usize::try_from(byte_index).unwrap();
```

Comment:
Confusing as hell, but it's in the CIP. I am not sure why it was settled on reading bits from the end. We have to document that properly in the Prelude.